### PR TITLE
Set the focus on Edit Box after edit action.

### LINKED ISSUE #3017: Right Click Popover -> Edit Comment does not focus on the Edit Comment Textbox
Coming from https://github.com/Expensify/Expensify/issues/163577

# Problem
If you right-click in the chat app to edit a comment, the edit comment textbox does not automatically get focused:
Kapture 2021-05-10 at 17 40 14

However this does not happen if you use the hover menu:
Kapture 2021-05-10 at 17 40 37

# Reproducible Steps
• Sign in to the e.chat staging web app
• Right-click text and choose Edit Comment
• Textbox does not automatically get focused

# Why is this important?
The two flows should be consistent.

# Solution
TBD

Upwork post - https://www.upwork.com/jobs/~015ee115217a359f1b

### LINKED ISSUE #3309: Edit Comment - After editing the comment, main compose box is not focused
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Main compose box should be focused so the user can start typing without needing to manually focus. 

## Actual Result:
Main compose box is not focused after editing a comment. 

## Action Performed:
1. Log in to e.cash 
2. Navigate to a conversation
3. Send a message. 
4. Right click on the message and click on edit comment
5. Edit the comment a bit and click enter

## Workaround:
User has to manually focus the compose box. 

## Platform:
Where is this issue occurring?

 Web ✔️ 
 iOS ✔️ 
 Android ✔️ 
 Desktop App (Unable to check) 
 Mobile Web ✔️ 

**Version Number:** 1.0.60-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856]

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/120554019-3c268e00-c3c7-11eb-8193-c43728ec188e.mp4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)
___
From @roryabraham https://expensify.slack.com/archives/C01GTK53T8Q/p1622578939426200
> After editing a message, the main ReportActionCompose is not refocused.

### PR DESCRIPTION
**Pull request by parasharrajat**
### Details
1. Remove modal visibility status from the Context Menu. which is needed to make the edited comment as a prime actor for focus.
2. Delayed the edit action after the Context menu is completely hidden which let us handle the focus. It removes the need for the SetTimeout hack.
3. Set the focus on the Edit box after each edit action.
4. Move the focus back to the main Composer when the Edit box is closed.

### Fixed Issues

Fixes #3017
fixes #3309 

### Tests | QA Steps (Web|Desktop)
* Sign in to the e.chat staging web app
* Right-click the text and choose Edit Comment
* Textbox should automatically get focused.
* Click Edit comment from Hover menu. Edit box should automatically get focused.
* Send a message, then hit the `ArrowUp` key. Edit box of your latest message should automatically get focused.
* Hit the `Esc` key, and the main `ReportActionCompose` should regain focus.

### Tests | QA Steps(mWeb|Android|IOS)
* Sign in to the e.chat staging web app
* Hold press any text message sent by you and choose Edit Comment
* Textbox should automatically get focused.

### Tests | QA Steps (Web|Desktop) for #3309
* Sign in to the e.chat staging web app
* Right-click the text and choose Edit Comment
* Textbox should automatically get focused.
* Click Edit comment from Hover menu. Edit box should automatically get focused.
* Save changes , main chat composer should regain focus.
* cancel Edit, main chat composer should regain focus.
* Click edit comment for any message.
* Hit the `Esc` key, and the main chat composer should regain focus.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [ ] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/24370807/121236427-93d76480-c8b3-11eb-9f20-9b8787a71e93.mp4

#### Mobile Web

https://user-images.githubusercontent.com/24370807/121238132-64c1f280-c8b5-11eb-868a-bbb082f05584.mp4

#### Desktop

https://user-images.githubusercontent.com/24370807/121236814-fa5c8280-c8b3-11eb-9c5b-4afd93dfcc4d.mp4

#### Android

https://user-images.githubusercontent.com/24370807/121239234-9ab3a680-c8b6-11eb-9209-a5ee0eaaade9.mp4

### CODE REVIEW

**Comment by parasharrajat** [comment]
Ready to roll. Thanks.

**Comment by roryabraham** [comment]
@parasharrajat Looks like lint is failing

**Comment by parasharrajat** [comment]
@roryabraham this is due to https://github.com/Expensify/Expensify.cash/pull/3401. And the fix is being made here https://github.com/Expensify/Expensify.cash/pull/3481. So I suggest we merge it quickly.

**Comment by roryabraham** [comment]
Lint error resolved, please merge `main` into this branch

**Comment by roryabraham** [comment]
Also @parasharrajat will this also fix [this deploy blocker](https://github.com/Expensify/Expensify.cash/pull/3434)?

**Comment by roryabraham** [comment]
> Also @parasharrajat will this also fix this deploy blocker?

Confirmed on iOS safari that it does.

**Review by roryabraham — CHANGES_REQUESTED** [review]
Also, I must be going crazy because I can't find where the `isFocused` prop of the `ReportActionCompose` comes from, because it isn't coming from [here](https://github.com/Expensify/Expensify.cash/blob/37465dbd07da1feab8347835d82ed3d2302cde4c/src/pages/home/report/ReportView.js#L37). Are we sure that this is ever being actually set? Can we safely get rid of it? We also have `isFocused` in the `ReportActionCompose` state.

File: src/pages/home/report/ReportActionContextMenu.js [hunk 1]
```diff
@@ -115,21 +116,24 @@ class ReportActionContextMenu extends React.Component {
                 icon: Pencil,
                 shouldShow: () => canEditReportAction(this.props.reportAction),
                 onPress: () => {
-                    this.hidePopover();
-                    saveReportActionDraft(
+                    const editAction = () => saveReportActionDraft(
                         this.props.reportID,
                         this.props.reportAction.reportActionID,
                         _.isEmpty(this.props.draftMessage) ? this.getActionText() : '',
                     );
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/pages/home/report/ReportActionContextMenu.js:123]: Yes

**Comment by parasharrajat** [comment]
`IsFocused` prop is coming from withNavigation HOC. It's the most important prop.  Removing it is not an option. :smile:

**Comment by parasharrajat** [comment]
Updated. Thanks.

**Comment by parasharrajat** [comment]
@deetergp @roryabraham Awaiting your feedback. Thanks.

**Comment by deetergp** [comment]
## Testing
### Web
Editing via the context menu always puts focus to the edit text box, but the right-click > edit only does it on the first edit after a page load. In subsequent right-click > edits, the focus stays at the bottom, in the main text editor.

### Desktop
Context menu Edit works like it should, but I couldn't get right-click > edit to focus in the edit text box, it always went down to the main text box.

### mWeb
Works as intended 👍 

### iOS
Works as intended, though it's weird that you only get the long-press context menu if you click on your own avatar. That's probably unrelated.

**Comment by parasharrajat** [comment]
@deetergp Thanks for letting me know. I changed the prop name but forgot to update it in the Component. `shouldSetModalVisibility` was the prop. Just fixed it. You can check now. Thanks.

**Comment by deetergp** [comment]
That fix did it, thanks @parasharrajat 👍

**Review by deetergp — APPROVED** [review]
Code looks good and tests pass.

**Comment by parasharrajat** [comment]
:shipit:

**Comment by parasharrajat** [comment]
@roryabraham Any feedback here? I think this PR is waiting to be merged by you. :sweat_smile:

**Review by roryabraham — APPROVED** [review]
This looks good to me. Sorry to have been a bottleneck here @parasharrajat – been a crazy couple days.

**Comment by kavimuru** [comment]
## iPhone - Safari -  Cursor focus not gained for edit comment field
## Expected Result: 
Textbox should automatically get focused.
## Actual Result:
Focus not gained in edit box
## Actions Performed:
1. Launch the URL and login
2. Hold press any text message sent by you and choose Edit Comment
## Platform:
iOS 
mWeb ✔️
Android 
## Build:
1.0.69-0

## Notes/Images/Video:

https://user-images.githubusercontent.com/43996225/122132148-50777b80-ce08-11eb-9be8-9bd8735cbc6f.mp4

**Comment by parasharrajat** [comment]
No worries I will test it on safari M-web

**Comment by roryabraham** [comment]
I am unable to reproduce on iOS 14.5 Safari. @isagoico @kavimuru can we please retest this in the latest staging version and see if it has been resolved?

**Comment by parasharrajat** [comment]
Same here issue is not reproducible on M-web safari iPhone 12 - 14.4.

**Comment by isagoico** [comment]
Retested and it was a pass in mWeb! Checking it off.

**Pull request merged**